### ISSUE TITLE
RTX 3080+ compatibility

### ISSUE THREAD

**Issue by catubc** [murthylab/sleap#454]
Hi
This is just a placeholder for an enhancement to have sleap work on RTX 3080 cards.  For a variety of reasons, RTX 3080 seem to be compatible only with tensorflow 2.4+ (and the link below suggests 2.5+ only as well).

https://github.com/tensorflow/tensorflow/issues/43701

I am happy to help test this.
Cat

**Comment by talmo** [comment]
Just an update/note for future reference:

Looked into this and the problem right now with updating to TF 2.4 are the CUDA dependencies (cudatoolkit=11.0 and cudnn=8.0.2).

While the [cudatoolkit](https://anaconda.org/anaconda/cudatoolkit/files) package is available on anaconda, [cudnn](https://anaconda.org/anaconda/cudnn/files) is not yet (latest is 7.6.5).

The update will also bump the numpy requirement to 1.19, which will affect a number of other dependencies. Assuming we can resolve those, one early route would be to provide an installation path without the conda-based CUDA packages and inform users that want to use this preview version that they have to install system-wide cudatoolkit and cudnn themselves.

Leaving this here in case other people run into this issue and would like to help us test it out.

**Comment by rlinus** [comment]
Hi 

I had the same issue with my RTX3090. I was able to solve it by using TF 2.2.2 instead of TF 2.3.1 (just change the requirement.txt file). You can use my dockerfile to test it:

https://github.com/rlinus/sleap/blob/develop/Dockerfile

**Comment by rlinus** [comment]
Hi 

I found a way to run SLEAP on my RTX3090. A docker container based on a tensorflow docker image from nvidia that comes with tensorflow 2.3.1 compiled with CUDA 11. You can test it with the following dockerfile that I made:

```
######################################
# Dockerfile for SLEAP GPU training  #
######################################

#Tensorflow v2.3.1 is needed for sleap v1.1.3
FROM nvcr.io/nvidia/tensorflow:20.11-tf2-py3

# Needed for sleap
RUN apt-get update && apt-get install -y libgl1
RUN python3 -m pip install sleap==1.1.3 --no-cache-dir

# Not needed, but can be useful
RUN apt-get update && apt-get install -y wget tree

# download test dataset https://sleap.ai/notebooks/Training_and_inference_on_an_example_dataset.html
RUN wget -O /root/dataset.zip https://github.com/murthylab/sleap-datasets/releases/download/dm-courtship-v1/drosophila-melanogaster-courtship.zip && mkdir /root/dataset && unzip /root/dataset.zip -d /root/dataset && rm /root/dataset.zip

WORKDIR /root

#Instructions:
#cd in folder with this file and then build the docker image with
#   docker build -t sleap .
#then run the image and open interactive shell with:
#   docker run --gpus all -it --rm sleap /bin/bash
# run this command in the container to test gpu:
#   sleap-train baseline.centroid.json "dataset/drosophila-melanogaster-courtship/courtship_labels.slp" --run_name "courtship.centroid" --video-paths "dataset/drosophila-melanogaster-courtship/20190128_113421.mp4"
```

you need a nvidia gpu enabled docker version for this to work.

best
linus

**Comment by talmo** [comment]
Thanks @rlinus! It might be handy to have a Docker container for future users.

In general this should work on any OS even without Docker if you manually install (the right versions of) CUDA + CuDNN.

@ariematsliah-princeton: Maybe we can add an automated docker build to our build workflows. This is important as people start to get their hands on 3080s and etc for which we currently do not have a supported installation path.

Since it's the same TensorFlow version, I imagine we won't have any API issues, but we should at least run the test suite to check against this particular build of TF.

@rlinus: we'd definitely welcome contributions or even reports if you run into any issues you might have with TF 2.3.1 + CUDA 11 :)

One more note for future reference: this probably won't work on Windows natively, but it might be possible with the new [GPU support in WSL2 + Docker](https://www.docker.com/blog/wsl-2-gpu-support-is-here/).

**Comment by catubc** [comment]
Thanks linus, I'm trying it now.
The issue for me is that I need to run a specific branch of sleap, namely the "supervised_id" branch. Is there a simple command/syntax to add to switch the branch?

**Comment by talmo** [comment]
@catubc: try this:
```
######################################
# Dockerfile for SLEAP GPU training  #
######################################

#Tensorflow v2.3.1 is needed for sleap v1.1.3
FROM nvcr.io/nvidia/tensorflow:20.11-tf2-py3

# Needed for sleap
RUN apt-get update && apt-get install -y libgl1
RUN python3 -m pip install git+https://github.com/murthylab/sleap.git@talmo/supervised_id --no-cache-dir

# Not needed, but can be useful
RUN apt-get update && apt-get install -y wget tree

# download test dataset https://sleap.ai/notebooks/Training_and_inference_on_an_example_dataset.html
RUN wget -O /root/dataset.zip https://github.com/murthylab/sleap-datasets/releases/download/dm-courtship-v1/drosophila-melanogaster-courtship.zip && mkdir /root/dataset && unzip /root/dataset.zip -d /root/dataset && rm /root/dataset.zip

WORKDIR /root
```

**Comment by rlinus** [comment]
@talmo just installing the right versions of CUDA + CuDNN is not enough, you would need to manually compile tensorflow with the right CUDA + CuDNN:
https://www.tensorflow.org/install/source

Running the container on Windows should be possible with WSL2:
https://www.docker.com/blog/wsl-2-gpu-support-is-here/

**Comment by catubc** [comment]
I confirm that the docker file updated for BottomUp ID works with RTX 3080. 

```
######################################
# Dockerfile for SLEAP GPU training  #
######################################

#Tensorflow v2.3.1 is needed for sleap v1.1.3
FROM nvcr.io/nvidia/tensorflow:20.11-tf2-py3

# Needed for sleap
RUN apt-get update && apt-get install -y libgl1
RUN python3 -m pip install git+https://github.com/murthylab/sleap.git@talmo/supervised_id --no-cache-dir

# Not needed, but can be useful
RUN apt-get update && apt-get install -y wget tree

# download test dataset https://sleap.ai/notebooks/Training_and_inference_on_an_example_dataset.html
RUN wget -O /root/dataset.zip https://github.com/murthylab/sleap-datasets/releases/download/dm-courtship-v1/drosophila-melanogaster-courtship.zip && mkdir /root/dataset && unzip /root/dataset.zip -d /root/dataset && rm /root/dataset.zip

WORKDIR /root

```

Some notes for any newbies like me in getting docker + GPU to work with sleap:

1. You will need nvidia docker to interface GPU to docker (otherwise you'll get errors when running the above). For Ubuntu this works:

sudo apt-get install -y nvidia-docker2

2.  Start the docker nvidia interface before mounting image

sudo systemctl restart docker

3. mounting the sleap docker image with (optinal) access to my Downloads directory requires a -v flag:

sudo docker run -v /media/cat/4TBSSD/Downloads:/mnt/mydata --gpus all -it --rm sleap /bin/bash

**Comment by catubc** [comment]
Hi 
So I've tested this some more, it seems my 3080 runs slower than  older GPUs using this docker option. 
But it works, and hopefully tf will be updated next few months to work properly with RTX 3rd gen.
Thanks!

**Comment by talmo** [comment]
Good news everyone! We now have a pre-release that should work on newer GPUs: https://github.com/murthylab/sleap/releases/tag/v1.2.0a0

Give it a go and please file more issues if things break!

**Comment by sommerc** [comment]
Hi Talmo,

I tried to install the pre-release on Win64 with conda by:

`conda create -n sleap_120 -c sleap sleap=1.2.0a0`

but I get:
```
PackagesNotFoundError: The following packages are not available from current channels:
  - sleap=1.2.0a0
```
I also tried to create a python=3.7 environment first and (conda) install from within, but with same result

Thanks a lot, SLEAP is amazing!
Chris

**Comment by talmo** [comment]
Sorry about that @sommerc! Here's the correct command to install the pre-release version:

`conda create -n sleap_120 -c sleap/label/dev sleap=1.2.0a0`

I'll update it in the release page as well. Let me know if you're still having any issues.

Cheers,

Talmo

**Comment by sommerc** [comment]
Thanks @talmo for the quick reply!

`conda create -n sleap_120 -c sleap -c sleap/label/dev sleap=1.2.0a0`

alone coplained about missing `tensorflow==2.7.0', but when I included the main sleap channel with

`conda create -n sleap_120 -c sleap -c sleap/label/dev -c sleap sleap=1.2.0a0`

it worked!

Starting the GUI with `sleap-label`, however, resulted in a version mismatch with the `wheel` package, which was resolved by

`conda install wheel==0.3.2`

Now, it's running and blazing fast on our new RTX A4000 :-)

Thanks again for everything!

Cheers,
Chris

**Comment by sunmart** [comment]
I'm running into the same issue with the wheel package, but I haven't been able to download wheel 0.3.2 Is that the right version? I tried to download 0.32.1 and that doesn't seem to work.

**Comment by talmo** [comment]
I believe this should work: `conda install wheel=0.32.0`

Let me know if it doesn't on your end. I'll work on a fix so this doesn't happen in the next release.

**Comment by sunmart** [comment]
Hi Talmo,

Thanks for the quick response, looks like 0.32.0 is not available and 0.32.1 requires older versions of python

(base) C:\Users\bambahmukkulab>conda install wheel==0.32.0
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: /
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

**Comment by talmo** [comment]
Hi @sunmart,

It looks like you're trying to install it in the `base` environment.

You'll want to `conda activate sleap_120` and then `conda install wheel=0.32.0`.

**Comment by sunmart** [comment]
Oh that fixed it. Thanks a ton!

Cheers,
Sunaina

**Comment by outlace** [comment]
I was also unable to get the official release of SLEAP to use my GPU (RTX A5000) but this pre-release version is working great so far!

**Comment by aperkes** [comment]
Installing the pre-release and fixing the wheel version also worked for me. I just bought a new computer with a geforce rtx 3060 12gb on Ubuntu 20.04. To get the gpu to work, I also needed to add the lib for the conda environment by adapting this code: https://stackoverflow.com/questions/46826497/conda-set-ld-library-path-for-env-only

Just create two files in their respective directories to temporarily update the LD_LIBRARY_PATH: 
`$ cat ~/anaconda3/envs/sleap/etc/conda/activate.d/activate.sh `
#!/bin/sh

ORIGINAL_LD_LIBRARY_PATH=$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=/home/ammon/anaconda3/envs/sleap/lib/:$LD_LIBRARY_PATH

`$ cat ~/anaconda3/envs/sleap/etc/conda/deactivate.d/deactivate.sh `
#!/bin/sh
export LD_LIBRARY_PATH=$ORIGINAL_LD_LIBRARY_PATH
unset ORIGINAL_LD_LIBRARY_PATH

This is especially helpful if you need multiple CUDA versions on your computer. In any case, now everything seems to be working.

**Comment by talmo** [comment]
Thanks @aperkes!

Is this still the case with the newest prerelease? → https://github.com/murthylab/sleap/releases/tag/v1.2.0a5

**Comment by aperkes** [comment]
I just tried with the newest pre-release. I no longer need to update wheel, however I still need the add the little activate script to update the path. I installed from conda (`conda create -y -n sleap -c sleap -c sleap/label/dev -c nvidia -c conda-forge sleap=1.2.0a5`) 

I suspect I did something sub-optimal when installing cuda (using sudo apt install nvidia-cuda-toolkit) or something, but it's working so I'm wary of changing something there and having to start over.

**Comment by talmo** [comment]
Hmm, yeah, I suspect this is related to https://github.com/tensorflow/tensorflow/issues/52988 -- and seems like there's not much we can do since tensorflow changed how it finds the cuda binaries in 2.7...

I think I might switch back to tensorflow 2.6 at least for the conda packages since I've run into this problem a couple of times now.

Thanks for the report @aperkes! Let me know if you run into any other issues, esp. GPU-related.

**Comment by talmo** [comment]
Hi all,

This should now be fixed in [SLEAP v1.2.0](https://github.com/murthylab/sleap/releases/tag/v1.2.0)! Head over to the [installation page](https://sleap.ai/installation.html) for instructions on setting it up.

Thanks for all the reports and workarounds. Please feel free to open a new issue if you're having trouble installing the new version.

Cheers,

Talmo
